### DISCUSSION TITLE
Data migration from confluence Kafka cluster to Strimzi kafka Cluster

### DISCUSSION THREAD

**Discussion by prateekkohli21** [strimzi/strimzi-kafka-operator#8053] [General]
I have a confluent kafka cluster installed on my kubernetes cluster. I need to upgrade it to Strimzi.

What are the ways to transfer data? I know using MM2 is an option, but is there a way to transfer all the data through pvc migration, i.e. I can just reuse the pvc created in confluent kafka and mount it inside Strimzi kafka.

**Comment by scholzj** [comment]
Mirror Maker would normally be the right option. 

Reusing the PVs would highly depend on what data they store on which path, what was the configuration, what permissions the data have, and so on. It might be possible if you can make the Confluent PVs look like the Strimzi PVs. Then you can just create new PVCs with the Strimzi names which would bind the PVs and re-use them.  But you would need to check how much they differ. It is certainly not something that works out of the box.

**Reply by prateekkohli21** [reply]
We will need to migrate zookeeper data as well in this case(if we try pvc reuse), right?

But if we use MM2 it would do all the migration. Is my understanding correct?

And also, would I be able to avoid data loss while using MM2 while the producers are continuously producing?

**Reply by scholzj** [reply]
With MM2, you do not do any special migration of the ZooKeeper data. If you want to copy the PVCs, you might need to. But you probably first want to make sure you use settings that are compatible with Strimzi. Also, keep in mind that Confluent Platform and Apache Kafka are not the same software. So there might be other differences as well. So while you can try the migration using the storage, there might be things you will need to figure out.

**Reply by prateekkohli21** [reply]
I might be able to migrate the pvc, the only problem I am currently facing is related to the mount path.
My older confluence pvc is mounted on /opt/kafka/data which has a directory named "logs" which has the kafka logs.

So, I reuse & mount that same pvc to /var/lib/kafka in Strimzi, but the only problem is that the old kafka data comes inside the "logs" folder, but Strimzi expects it to be in "kafka-log${broker-id}" directory. 

Would it be safe to change Strimzi logs directory to /var/lib/kafka/logs in KafkaBrokerConfigurationBuilder.java class, or can it have any other adverse effects, because I can also see /var/lib/kafka/kafka-log0 directory being used in Capacity.java class for CruiseControl in JBOD storage.

Can I change it or is there any other solution by which I can mount opt/kafka/data/logs from old pvc to /var/lib/kafka/kafka-logs0 in new pvc?

**Reply by scholzj** [reply]
If you change the path in `KafkaBrokerConfigurationBuilder`, you will be forever left to your own fork of Strimzi. So I think the best way is to edit the PVs and change the directory structure first.

**Reply by prateekkohli21** [reply]
Thanks for the response!!

Will try to edit pv, but if that doesn't work for my case, would it be safe to still change the path KafkaBrokerConfigurationBuilder.java, as I can maintain my own fork for Strimzi.
Does changing this path have any other dependency as well?

**Reply by scholzj** [reply]
> Will try to edit pv, but if that doesn't work for my case, would it be safe to still change the path KafkaBrokerConfigurationBuilder.java, as I can maintain my own fork for Strimzi.

Your call ... but I think it would not be a good choice.

> Does changing this path have any other dependency as well?

No idea, I don't think anyone ever did that.

**Reply by prateekkohli21** [reply]
Sorry I am a bit new to Kubernetes. Did not find any way to edit the directory structure in a PV.
Would you be able to help on that?

**Reply by scholzj** [reply]
You would basically need to create some helper pod which mounts the PVC / PV. And then in there you would need to update the structure (rename the directory, update the permissions, etc. - what ever might be needed.)

**Reply by prateekkohli21** [reply]
Thanks a lot, will surely try to explore that.

Just 1 last question, why is the pvc name different in case of JBOD storage, the jbod storage id is appended (data-${jbod-storage-id})? Is there any reason to append the storage id in case of JBOS storage.

**Reply by scholzj** [reply]
The JBOD storage allows you to use multiple disks. So the disk ID has to be reflected in the name.

**Reply by prateekkohli21** [reply]
So, it is just for identification. Is my understanding correct?

**Reply by scholzj** [reply]
I'm not sure how exactly you mean with `just for the identifcation` TBH. It is _just for identification_, yes. But you need to follow the naming exactly to make it work.

**Reply by prateekkohli21** [reply]
Thanks for clearing the doubts!!
